lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -411,6 +417,29 @@ collectOpsToPipeline(scf::ForOp forOp,
     opInfo[loadOp] = loadInfo;
   }
 
+  // A loadOp may not have valid stage info, e.g. for cases where the loaded
```

Comment:
How do we end up with a load that is not on a chain that leads to the dot on the list? I would think that `loadOpsToDistanceAndUse` should produce only such loads, right?

Reply:
Thanks for the quick review, @pawelszczerbuk! Let's take the test kernel ```matmul_indirect_pipeline``` added along with this PR as the running example, where we have (I am keeping only the relevant pieces):

```
    scf.for %arg4 = %c0_i32 to %c2_i32 step %c1_i32  : i32 {
      %15 = tt.load %13
      %16 = tt.addptr %14, %15
      %17 = tt.load %16
      %18 = tt.expand_dims %17 {axis = 0 : i32}
      %19 = tt.broadcast %18
      %20 = arith.addf %9, %19
      %21 = triton_gpu.convert_layout %9
      %22 = triton_gpu.convert_layout %20
      %23 = tt.dot %21, %22, %cst
```

With this input, ```loadOpsToDistanceAndUse``` gives us the following ```loadOpToDistAndUse```:

```
# loadOp -> (dist, use)
tt.load %16 -> (0, tt.dot %21, %22, %cst)
tt.load %13 -> (1, tt.load %16)
```

When we update the opInfo in the for loop below:

```
  for (auto &[loadOp, distAndUse] : loadOpToDistAndUse) {
    ...
    bool isLoadDotOperand = loadDotOperand(loadOp, loadIsMMAV3, sharedEnc);
    ...
  }
```

When we handle the ```tt.load %16``` item, ```loadDotOperand``` returns false because the user of ```tt.load %16``` is not a ```tt.dot``` nor a ```triton_gpu.convert_layout```. We end up with the relevant stage value being -1 for ```tt.load %16```. This eventually triggers the assertion failure at:

```
    if (iter != opInfo.end() && iter->second.stage == -1) {
        assert(!isa<tt::LoadOp>(iter->first));
```

We might be able to avoid generating ```tt.load %16 -> (0, tt.dot %21, %22, %cst)``` mapping in ```loadOpToDistAndUse``` by adding more constraints. On the other hand, I feel that those constraints might look similar to what we do in ```loadDotOperand```. Leaving the code as it whiling adding an extra routine to remove invalid loadOp(s) seems to be cleaner. 

The IR above was generated from the following kernel, which was reduced from a complex Triton kernel. The original Triton kernel used to work without assertion failures:

```
import triton
import triton.language as tl

@triton.jit
def foo(
    in_ptrs,
    ts_ptr1,
    ts_ptr2,
    out_ptrs,
):
    offs = tl.arange(0, 32)
    off_o = offs[:, None] + offs[None, :]
    a = tl.load(in_ptrs + off_o)
    out_ptrs = out_ptrs + off_o

    for _ in range(0, 2):
        ts_1 = tl.load(ts_ptr1 + offs)
        ts_2 = tl.load(ts_ptr2 + ts_1)
        b = a + ts_2
        x = tl.dot(a, b)
        tl.store(out_ptrs, x)
```

Reply:
Oh, I see now where the issue is coming from! Can you see if it would be still present after this change?
github.com/openai/triton/pull/3415/
It changes the logic around the `loadDotOperand`, and I think it might actually make it consistent, so we should not hit this case (fingers crossed)

Reply:
The commit from Justin fixed the failure for the attached ttgir, but introduced some other issues when I tried it with the original Triton kernel. I will get you back with more details later. Thanks!

Reply:
Follow up on this:

https://github.com/openai/triton/issues/3468
https://github.com/openai/triton/pull/3469